Introduce SubframePageProxy and get FrameInfoData from a WebFrameProxy's current process

achristensen07 opened the pull request:
#### 377a279a5f9addf4042163a8abaf1cca0396bdae
<pre>
Introduce SubframePageProxy and get FrameInfoData from a WebFrameProxy&apos;s current process
<a href="https://bugs.webkit.org/show_bug.cgi?id=249522">https://bugs.webkit.org/show_bug.cgi?id=249522</a>
rdar://103475229

Reviewed by Chris Dumez.

SubframePageProxy is the receiver of messages from a WebPage containing a WebFrame that has
a parent frame in another process.

Rather than get all the FrameInfoData from one process, send messages to whatever processes
a WebFrameProxy is using and aggregate the data in the UI process.  To do this, change the
child frames container from a HashSet to a ListHashSet to match the deterministic ordering
that currently exists in the web process.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _frames:]):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
* Source/WebKit/UIProcess/SubframePageProxy.cpp: Added.
(WebKit::SubframePageProxy::SubframePageProxy):
(WebKit::SubframePageProxy::~SubframePageProxy):
(WebKit::SubframePageProxy::messageSenderConnection const):
(WebKit::SubframePageProxy::messageSenderDestinationID const):
(WebKit::SubframePageProxy::didReceiveMessage):
(WebKit::SubframePageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/SubframePageProxy.h: Added.
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::~WebFrameProxy):
(WebKit::WebFrameProxy::didFinishLoad):
(WebKit::WebFrameProxy::commitProvisionalFrame):
(WebKit::WebFrameProxy::getFrameInfo):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getAllFrames):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::getFrameInfo):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::frameTreeNodeData): Deleted.
(WebKit::WebPage::getAllFrames): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/258149@main">https://commits.webkit.org/258149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd18cfee840ec1f55789f6ba0142707afec2cabe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100939 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110243 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170510 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/963 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93350 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108083 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34946 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90248 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22994 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77925 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3774 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24512 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3795 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/912 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9907 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44014 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5570 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2933 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1977,28 +1977,6 @@ WebPage& WebPage::fromCorePage(Page& page)
-    std::optional<FrameTreeNodeData> info;
-    if (auto* webFrame = WebFrame::fromCoreFrame(frame)) {
-        Vector<FrameTreeNodeData> children;
-        for (auto* childFrame = frame.tree().firstChild(); childFrame; childFrame = childFrame->tree().nextSibling()) {
```
So we used to get frame info data in tree order. However, after your change, I don't believe this is the case anymore. Although you switched to a ListHashSet to get some sort of a deterministic order, this is not the more logical tree order, which is very unfortunate.

I think the first step should be to have a FrameTree representation in the UIProcess (which we will need for process isolation no matter what). Then we could maintain tree order while re-implementing this.

achristensen07 replied:
FrameTree::appendChild is the only way to add to a FrameTree right now, so creation order is the same order as what we had before.

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -400,13 +400,51 @@ void WebFrameProxy::commitProvisionalFrame(FrameIdentifier frameID, FrameInfoDat
+        static Ref<FrameInfoCallbackAggregator> create(CompletionHandler<void(std::optional<FrameTreeNodeData>&&)>&& completionHandler) { return adoptRef(*new FrameInfoCallbackAggregator(WTFMove(completionHandler))); }
+        void setCurrentFrameData(FrameInfoData&& data) { m_currentFrameData = WTFMove(data); }
+        void addChildFrameData(FrameTreeNodeData&& data) { m_childFrameData.append(WTFMove(data)); }
+        FrameInfoCallbackAggregator(CompletionHandler<void(std::optional<FrameTreeNodeData>&&)>&& completionHandler)
```
This shouldn't be public.

achristensen07 committed:
Introduce SubframePageProxy and get FrameInfoData from a WebFrameProxy's current process
https://bugs.webkit.org/show_bug.cgi?id=249522
rdar://103475229

Reviewed by Chris Dumez.

SubframePageProxy is the receiver of messages from a WebPage containing a WebFrame that has
a parent frame in another process.

Rather than get all the FrameInfoData from one process, send messages to whatever processes
a WebFrameProxy is using and aggregate the data in the UI process.  To do this, change the
child frames container from a HashSet to a ListHashSet to match the deterministic ordering
that currently exists in the web process.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _frames:]):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
* Source/WebKit/UIProcess/SubframePageProxy.cpp: Added.
(WebKit::SubframePageProxy::SubframePageProxy):
(WebKit::SubframePageProxy::~SubframePageProxy):
(WebKit::SubframePageProxy::messageSenderConnection const):
(WebKit::SubframePageProxy::messageSenderDestinationID const):
(WebKit::SubframePageProxy::didReceiveMessage):
(WebKit::SubframePageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/SubframePageProxy.h: Added.
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::~WebFrameProxy):
(WebKit::WebFrameProxy::didFinishLoad):
(WebKit::WebFrameProxy::commitProvisionalFrame):
(WebKit::WebFrameProxy::getFrameInfo):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getAllFrames):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::getFrameInfo):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::frameTreeNodeData): Deleted.
(WebKit::WebPage::getAllFrames): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/258149@main
